Setup storage

thewrongjames opened the pull request:
This creates a useStorage hook that the root app component uses to gain access to a storage object that is synchronised with the browser localStorage object to allow persistence of data across sessions. This object is not designed to be used by child components, but instead to be passed to contexts that use it as the state-owner for various particular use cases. The only existing one of these at the moment is the theme context, so the functional upshot of this change is that the theme is now remembered.

There is a lot of code here for saving/serialising and type-safely loading/deserialising from localStorage. Not sure this is the way we will want to do it forever (especially if we look into robustly versioning the schema for the data we store in local storage), but it seems nice-ish and works for now.

This also has a bunch of other changes to themes that came up as I was working on this (for instance, hopefully now actually calling them themes everywhere rather than colour schemes as they were called before), as the changes we reasonable coupled to that.

thewrongjames committed:
Setup a storage context

- This is very much a first pass, and I'm going to change things.
- The storage context doesn't currently use local storage, it just has a
  reducer state in memory, but this was just to get the setup working,
  with the idea being that the storage system should be able to hide
  however it actually persists memory from whatever uses it.
- I set up the theme context to use the storage context.
- I did a lot of wrangling in src/index.tsx for this. More than I
  probably should to make arbitrary application nice.
- In light of this, I think storage shouldn't actually be a context. It
  should just be something that other contexts use, but I wanted to
  commit this version so I don't lose it.

thewrongjames committed:
Switch storage to just being a reducer

It will probably need to switch to being a hook once I add the side
effect of actually persisting the state.

But, this greatly reduces the complexity of logic that has to happen at
the top level.

thewrongjames committed:
Switch storage to be a hook

Though at the moment it doesn't do anything more than useReducer.

thewrongjames committed:
Add live-updating of the system theme

Possibly not doesn't belong on this branch in retrospect but there it
is.

thewrongjames committed:
Actually store the storage in local storage

Most of the actual code here though was in validating and serialising
and deserialisng and whatnot. Not sure that we will stick with this
forever, and there are a lot of todos, but I like this for now.

moveitroach commented on src/models/storage-data/index.tsx:
```diff
@@ -0,0 +1,46 @@
+    if (!(error instanceof TypeParsingError)) {
+      throw error
+    }
+
```
maybe we should log the type parsing error here since you already wrote those out

moveitroach reviewed (approved):
there's one thing you can add but it's not critical

thewrongjames committed:
Log errors on failed storage deserialisation

A better approach will probably be needed in the long run still. The
deserialisation function should probably error and be handled at the
highest level.
